Index: MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c
===================================================================
--- MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c	(revision 11711)
+++ MdeModulePkg/Universal/Acpi/AcpiTableDxe/AcpiTableProtocol.c	(working copy)
@@ -1,6 +1,8 @@
 /** @file
   ACPI Table Protocol Implementation
 
+  Modified to support another ACPI Table, such as Xen ACPI Table. - Bei Guan <gbtju85@gmail.com>
+
   Copyright (c) 2006 - 2011, Intel Corporation. All rights reserved.<BR>
   This program and the accompanying materials
   are licensed and made available under the terms and conditions of the BSD License
@@ -202,6 +204,32 @@
 }
 
 /**
+  Function finds a system configuration table that is stored in the
+  EFI System Table based on the provided GUID.
+
+  @param[in]  TableGuid    A pointer to the table's GUID type.
+
+  @retval EFI_SUCCESS      A configuration table matching TableGuid was found.
+  @retval EFI_NOT_FOUND    A configuration table matching TableGuid was not found.
+**/
+EFI_STATUS
+EFIAPI
+GetSystemConfigurationTable (
+  IN EFI_GUID *TableGuid
+  )
+{
+  UINTN Index;
+
+  for (Index = 0; Index < gST->NumberOfTableEntries; Index++) {
+    if (CompareGuid (TableGuid, &(gST->ConfigurationTable[Index].VendorGuid))) {
+      return EFI_SUCCESS;
+    }
+  }
+
+  return EFI_NOT_FOUND;
+}
+
+/**
   This function publishes the specified versions of the ACPI tables by
   installing EFI configuration table entries for them.  Any combination of
   table versions can be published.
@@ -226,6 +254,19 @@
   UINT64                    Buffer64;
 
   //
+  // See if there is already a ACPI table in System Configure table
+  //
+  Status = GetSystemConfigurationTable (&gEfiAcpi10TableGuid);
+  if (Status == EFI_SUCCESS) {
+    AcpiTableInstance->TablesInstalled1 = TRUE;
+  }
+
+  Status = GetSystemConfigurationTable (&gEfiAcpiTableGuid);
+  if (Status == EFI_SUCCESS) {
+    AcpiTableInstance->TablesInstalled3 = TRUE;
+  }
+
+  //
   // Reorder tables as some operating systems don't seem to find the
   // FADT correctly if it is not in the first few entries
   //
