diff -r 56a3b9c7367f tools/libxl/libxl.idl
--- a/tools/libxl/libxl.idl	Tue Apr 12 14:55:25 2011 +0100
+++ b/tools/libxl/libxl.idl	Thu Jul 07 00:12:44 2011 +0800
@@ -75,6 +75,7 @@
                                              
 libxl_domain_create_info = Struct("domain_create_info",[
     ("hvm",          bool),
+    ("hvmbios",      string),
     ("hap",          bool),
     ("oos",          bool),
     ("ssidref",      integer),
diff -r 56a3b9c7367f tools/libxl/libxl_create.c
--- a/tools/libxl/libxl_create.c	Tue Apr 12 14:55:25 2011 +0100
+++ b/tools/libxl/libxl_create.c	Thu Jul 07 00:12:44 2011 +0800
@@ -371,6 +371,10 @@
     if (info->poolname)
         xs_write(ctx->xsh, t, libxl__sprintf(gc, "%s/pool_name", vm_path), info->poolname, strlen(info->poolname));
 
+    if (info->hvmbios){
+        xs_write(ctx->xsh, t, libxl__sprintf(gc, "%s/hvmloader/bios", dom_path), info->hvmbios, strlen(info->hvmbios));
+    }
+
     libxl__xs_writev(gc, t, dom_path, info->xsdata);
     libxl__xs_writev(gc, t, libxl__sprintf(gc, "%s/platform", dom_path), info->platformdata);
 
diff -r 56a3b9c7367f tools/libxl/xl_cmdimpl.c
--- a/tools/libxl/xl_cmdimpl.c	Tue Apr 12 14:55:25 2011 +0100
+++ b/tools/libxl/xl_cmdimpl.c	Thu Jul 07 00:12:44 2011 +0800
@@ -642,6 +642,10 @@
 
     libxl_init_create_info(c_info);
 
+    if (!xlu_cfg_get_string (config, "hvmbios", &buf)){
+        c_info->hvmbios = strdup(buf);
+    }
+
     c_info->hvm = 0;
     if (!xlu_cfg_get_string (config, "builder", &buf) &&
         !strncmp(buf, "hvm", strlen(buf)))
